### PR TITLE
Add CNTR and WAIT support for rdm_mbw_mr

### LINKED ISSUE #51: run_cray_tests.py runs rdm_mbw_mr with invalid option causing -nan result but still counts as a pass
On jupiter, running `run_cray_tests.py`, the following output was observed for `rdm_mbw_mr`:
```
jpetro@jupiter:~/libfabric/install/bin> ./run_cray_tests.py -p gni -l srun
...
Starting: srun --nodes=2 --exclusive -t00:10:00 --ntasks-per-node=2 --cpus-per-task=1 --cpu_bind=none ./rdm_mbw_mr -pgni
###
### srun --nodes=2 --exclusive -t00:10:00 --ntasks-per-node=2 --cpus-per-task=1 --cpu_bind=none ./rdm_mbw_mr -pgni
###
# Libfabric Multiple Bandwidth and Message Rate Test
# fi_send  -> fi_recv
# [ pairs: 0 ] [ window size: 64 ]
# Size               Iov count                MB/s          Messages/s
1                            1                -nan                -nan
2                            2                -nan                -nan
3                            3                -nan                -nan
4                            4                -nan                -nan
5                            5                -nan                -nan
6                            6                -nan                -nan
7                            7                -nan                -nan
8                            8                -nan                -nan
...
# Libfabric Multiple Bandwidth and Message Rate Test
# fi_sendv -> fi_recv  (Scatter/Gather)
# [ pairs: 0 ] [ window size: 64 ]
# Size               Iov count                MB/s          Messages/s
1                            1                -nan                -nan
2                            2                -nan                -nan
3                            3                -nan                -nan
4                            4                -nan                -nan
5                            5                -nan                -nan
6                            6                -nan                -nan
7                            7                -nan                -nan
8                            8                -nan                -nan

2                            1                -nan                -nan
4                            2                -nan                -nan
6                            3                -nan                -nan
8                            4                -nan                -nan
10                           5                -nan                -nan
12                           6                -nan                -nan
14                           7                -nan                -nan
16                           8                -nan                -nan
...
###
### EXIT STATUS: 0
###
```
This is also counted as a pass. 
```
###
### 14 PASSSED
### 0 FAILED
### 0 NOT RUN
```
Running the help message for `rdm_mbw_mr` shows that there are two `-p` options.
```
jpetro@jupiter:~/libfabric/install/bin> srun ./rdm_mbw_mr -h

Libfabric Multiple Bandwidth and Message Rate Test

Options:
 -r <0,1>             Print uni-directional message rate (default 1)
 -p <pairs>           Number of pairs involved (default np / 2)
 -w <window>          Number of messages sent before acknowldgement (64, 10) [cannot be used with -v]
 -v                   Vary the window size (default no) [cannot be used with -w]
 -p <provider>        specified provider (e.g., gni)
```
`run_cray_tests.py` uses the `-pgni` option which I suspect is setting pairs to 0 on a failed `atoi()` call.  

The test runs normally (as far as I can tell) if I run it by hand with no `-p` option.

### PR DESCRIPTION
**Pull request by jshimek**
- Change -p <pairs> to -x so as to not overload the -p <provider>
  command line check.
- Add -m <wait_type> so different wait options can be tested.
	Options include 0(FI_WAIT_NONE), 1(FI_WAIT_UNSPEC), 2(FI_WAIT_SET)
- Add -c <completion_type> so different completion objects can be tested.
	Options include 0(CQ), 1(CNTR)

Signed-off-by: James Shimek <jshimek@cray.com>

Fixes #51

### CODE REVIEW

File: performance/multi-node/rdm_mbw_mr.c [hunk 1]
```diff
@@ -148,52 +169,117 @@ static void cq_readerr(struct fid_cq *cq, const char *cq_str)
...
+				ret = fi_cntr_wait(cntr, num_completions, -1);
+		}
+
+		if (wait_type == FI_WAIT_SET) {
+			ret = fi_wait(waitset, -1);
+		}
+
+		if (ret < 0 && (wait_type == FI_WAIT_SET ||
+				wait_type == FI_WAIT_UNSPEC)) {
+			fprintf(stdout, "Waiting timed out %d\n", ret);
```
Comment (chuckfossen) [inline performance/multi-node/rdm_mbw_mr.c:216]: If you time out, are you going to break out of the loop or continue calling wait?

File: performance/multi-node/rdm_mbw_mr.c [hunk 2]
```diff
@@ -536,18 +647,34 @@ int main(int argc, char *argv[])
...
 		return -1;
 
-	while ((op = getopt(argc, argv, "hp:w:vr:" CT_STD_OPTS)) != -1) {
+	while ((op = getopt(argc, argv, "hx:p:w:vr:m:c:" CT_STD_OPTS)) != -1) {
 		switch (op) {
 		default:
 			ct_parse_std_opts(op, optarg, hints);
 			break;
-		case 'p':
+		case 'x':
```
Comment (chuckfossen) [inline performance/multi-node/rdm_mbw_mr.c:656]: Why change p to x?

File: performance/multi-node/rdm_mbw_mr.c [hunk 3]
```diff
@@ -571,6 +698,24 @@ int main(int argc, char *argv[])
...
+		if (test_type == CNTR_TEST)
+			fprintf(stderr, "Test will use CNTR Completions\n");
+
+		if (wait_type == FI_WAIT_NONE)
+			fprintf(stderr, "WaitObject type is FI_WAIT_NONE\n");
+
+		if (wait_type == FI_WAIT_UNSPEC)
+			fprintf(stderr, "WaitObject type is FI_WAIT_UNSPEC\n");
+
+		if (wait_type == FI_WAIT_SET)
```
Comment (chuckfossen) [inline performance/multi-node/rdm_mbw_mr.c]: A switch statement seems appropriate here.

File: performance/multi-node/rdm_mbw_mr.c (hunk 2 above)
Comment (jshimek) [reply performance/multi-node/rdm_mbw_mr.c:656]: p is already the standard for change the provider type, and when you tried to use -p sockets the test would use sockets but then set pairs to something undefined. Switching to x allows the ct_parse_std_opts to work and the pair argument to work.

File: performance/multi-node/rdm_mbw_mr.c (hunk 1 above)
Comment (jshimek) [reply performance/multi-node/rdm_mbw_mr.c:216]: So it shouldn't ever timeout due to the usage of -1 to fi_wait/fi_cntr_wait, so the refactor keeps essentially the same exit conditions as the original, if wanted I could also add a timeout control variable to this to allow the user to change the default timeout for the fi_cntr_wait and fi_wait options.

File: performance/multi-node/rdm_mbw_mr.c (hunk 3 above)
Comment (jshimek) [reply performance/multi-node/rdm_mbw_mr.c]: In my haste to clean this up, that thought did not occur to me. I had originally located all these fprintf's in different locations, but moved them all here. I'll change this to a switch when I address any other needed changes.

**Comment by sungeunchoi** [comment]
Can you add some of these to the run script?  Is it reasonable to run all variations?

**Comment by jshimek** [comment]
The base test with default options takes a while to run, I wasn't sure we wanted to increase the runtime by 6x to cover all the different options available.

**Review by sungeunchoi — APPROVED** [review]

**Comment by sungeunchoi** [comment]
@jshimek will open a new issue for script changes.

**Pull request merged**